src/Cli/dotnet/PerformanceLogEventListener.cs

```diff
@@ -0,0 +1,165 @@
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Extensions.EnvironmentAbstractions;
+
+namespace Microsoft.DotNet.Cli.Utils
+{
+    internal sealed class PerformanceLogEventListener : EventListener
+    {
+        internal struct ProviderConfiguration
+        {
+            internal string Name { get; set; }
+            internal EventKeywords Keywords { get; set; }
+            internal EventLevel Level { get; set; }
+        }
+
+        private static ProviderConfiguration[] s_config = new ProviderConfiguration[]
+        {
+            new ProviderConfiguration()
+            {
+                Name = "Microsoft-Dotnet-CLI-Performance",
+                Keywords = EventKeywords.All,
+                Level = EventLevel.Verbose
+            }
+        };
+
+        private const char EventDelimiter = '\n';
+        private string _processIDStr;
+        private StreamWriter _writer;
+
+        [ThreadStatic]
+        private StringBuilder s_builder;
+
+        internal static PerformanceLogEventListener Create(IFileSystem fileSystem, string logDirectory)
+        {
+            // Only create a listener if the log directory exists.
+            if(string.IsNullOrWhiteSpace(logDirectory) || !fileSystem.Directory.Exists(logDirectory))
+            {
+                return null;
+            }
+
+            PerformanceLogEventListener eventListener = null;
+            try
+            {
+                // Initialization happens as a separate step and not in the constructor to ensure that
+                // if an exception is thrown during init, we have the opportunity to dispose of the listener,
+                // which will disable any EventSources that have been enabled.  Any EventSources that existed before
+                // this EventListener will be passed to OnEventSourceCreated before our constructor is called, so
+                // we if we do this work in the constructor, and don't get an opportunity to call Dispose, the
+                // EventSources will remain enabled even if there aren't any consuming EventListeners.
+                eventListener = new PerformanceLogEventListener();
+                eventListener.Initialize(fileSystem, logDirectory);
+            }
+            catch
+            {
+                if(eventListener != null)
+                {
+                    eventListener.Dispose();
+                }
+            }
+
+            return eventListener;
+        }
+
+        private PerformanceLogEventListener()
+        {
+        }
+
+        internal void Initialize(IFileSystem fileSystem, string logDirectory)
+        {
+            // TODO: Should we choose something that is guaranteed to be unique?
+            _processIDStr = Process.GetCurrentProcess().Id.ToString();
+            string logFilePath = Path.Combine(logDirectory, $"perf-{_processIDStr}.log");
```

Comment:
We have https://github.com/dotnet/sdk/blob/b1223209644d900702287faea8e9b71f95ec49f8/src/Cli/dotnet/Telemetry/PersistenceChannel/StorageService.cs  (copied from Hockeyapp SDK) pretty much doing the samething. We should try to reuse it

Reply:
The file you referenced does look similar.  Considering using it does seem reasonable, but we would also need to make sure that the additional functionality required by the perf log is there (e.g. treating individual logs as directories of files rather than individual files as telemetry does).

---
In reply to: [505751742](https://github.com/wli3/sdk/pull/1#discussion_r505751742) [](ancestors = 505751742)